src/rgw/rgw_admin.cc

```diff
@@ -1061,6146 +67,1088 @@ class StoreDestructor {
-    } else {
-      string bucket_instance_id = bucket_name + ":" + bucket_id;
-      r = store->get_bucket_instance_info(obj_ctx, bucket_instance_id, bucket_info, NULL, pattrs);
-    }
-    if (r < 0) {
-      cerr << "could not get bucket info for bucket=" << bucket_name << std::endl;
-      return r;
-    }
-    bucket = bucket_info.bucket;
-  }
-  return 0;
-}
-
-static int read_input(const string& infile, bufferlist& bl)
+#ifdef BUILDING_FOR_EMBEDDED
+extern "C" int cephd_rgw_admin(int argc, const char **argv)
+#else
+int main(int argc, const char **argv)
+#endif
 {
-  int fd = 0;
-  if (infile.size()) {
-    fd = open(infile.c_str(), O_RDONLY);
-    if (fd < 0) {
-      int err = -errno;
-      cerr << "error reading input file " << infile << std::endl;
-      return err;
-    }
-  }
-
-#define READ_CHUNK 8196
-  int r;
-  int err;
-
-  do {
-    char buf[READ_CHUNK];
-
-    r = safe_read(fd, buf, READ_CHUNK);
-    if (r < 0) {
-      err = -errno;
-      cerr << "error while reading input" << std::endl;
-      goto out;
-    }
-    bl.append(buf, r);
-  } while (r > 0);
-  err = 0;
-
- out:
-  if (infile.size()) {
-    close(fd);
-  }
-  return err;
-}
+  vector<const char*> args;
+  argv_to_vec(argc, (const char **)argv, args);
+  env_to_vec(args);
 
-template <class T>
-static int read_decode_json(const string& infile, T& t)
-{
-  bufferlist bl;
-  int ret = read_input(infile, bl);
-  if (ret < 0) {
-    cerr << "ERROR: failed to read input: " << cpp_strerror(-ret) << std::endl;
-    return ret;
-  }
-  JSONParser p;
-  if (!p.parse(bl.c_str(), bl.length())) {
-    cout << "failed to parse JSON" << std::endl;
-    return -EINVAL;
-  }
+  boost::intrusive_ptr<CephContext> cct = global_init(nullptr, args, CEPH_ENTITY_TYPE_CLIENT,
+                                                      CODE_ENVIRONMENT_UTILITY, 0);
 
-  try {
-    decode_json_obj(t, &p);
-  } catch (JSONDecoder::err& e) {
-    cout << "failed to decode JSON input: " << e.message << std::endl;
-    return -EINVAL;
+  // for region -> zonegroup conversion (must happen before common_init_finish())
+  if (!g_conf->rgw_region.empty() && g_conf->rgw_zonegroup.empty()) {
+    g_conf->set_val_or_die("rgw_zonegroup", g_conf->rgw_region.c_str());
   }
-  return 0;
-}
 
-template <class T, class K>
-static int read_decode_json(const string& infile, T& t, K *k)
-{
-  bufferlist bl;
-  int ret = read_input(infile, bl);
-  if (ret < 0) {
-    cerr << "ERROR: failed to read input: " << cpp_strerror(-ret) << std::endl;
-    return ret;
-  }
-  JSONParser p;
-  if (!p.parse(bl.c_str(), bl.length())) {
-    cout << "failed to parse JSON" << std::endl;
-    return -EINVAL;
-  }
+  common_init_finish(g_ceph_context);
 
-  try {
-    t.decode_json(&p, k);
-  } catch (JSONDecoder::err& e) {
-    cout << "failed to decode JSON input: " << e.message << std::endl;
-    return -EINVAL;
-  }
-  return 0;
-}
+  rgw_user user_id;
```

Comment:
two options:
1. create a new struct with those variables and pass it to the parsing functions. 
2. make those variables global and avoid passing them to the paring functions.

improvement for the future is using boost:optional